source/extensions/filters/http/grpc_shim/grpc_shim.cc

```diff
@@ -0,0 +1,145 @@
+#include "extensions/filters/http/grpc_shim/grpc_shim.h"
+
+#include <netinet/in.h>
+
+#include "envoy/common/platform.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/grpc/common.h"
+#include "common/grpc/status.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace {
+Envoy::Grpc::Status::GrpcStatus grpcStatusFromHeaders(Envoy::Http::HeaderMap& headers) {
+  const auto http_response_status = Envoy::Http::Utility::getResponseStatus(headers);
+
+  // Notably, we treat an upstream 200 as a successful response. This differs
+  // from the standard but is key in being able to transform a successful
+  // upstream HTTP response into a gRPC response.
+  if (http_response_status == 200) {
+    return Envoy::Grpc::Status::GrpcStatus::Ok;
+  } else {
+    return Envoy::Grpc::Utility::httpToGrpcStatus(http_response_status);
+  }
+}
+} // namespace
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace GrpcShim {
+
+Http::FilterHeadersStatus GrpcShim::decodeHeaders(Http::HeaderMap& headers, bool end_stream) {
+  // Short circuit if header only.
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // TODO(snowp): Add an enabled flag so that this filter can be enabled on a per route basis.
+
+  // If this is a gRPC request we:
+  //  - mark this request as being gRPC
+  //  - change the content-type to application/x-protobuf
+  if (Envoy::Grpc::Common::hasGrpcContentType(headers)) {
+    enabled_ = true;
+
+    // We keep track of the original content-type to ensure that we handle
+    // gRPC content type variations such as application/grpc+proto.
+    content_type_ = headers.ContentType()->value().c_str();
+    headers.ContentType()->value(upstream_content_type_);
+    headers.insertAccept().value(upstream_content_type_);
+
+    // Clear the route cache to recompute the cache. This provides additional
+    // flexibility around request modification through the route table.
+    decoder_callbacks_->clearRouteCache();
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcShim::decodeData(Buffer::Instance& buffer, bool) {
+  // If this is a gRPC request and this is the start of DATA, remove the first 5
+  // bytes. These correspond to the gRPC data header.
+  if (enabled_ && !prefix_stripped_) {
+    buffer.drain(5);
```

Comment:
This will crash with malicious downstream when buffer is < 5 bytes length.

Btw, aren't this messing up Content-Length in header?

Reply:
Ah yeah, you're right. Should be easy enough to +/- 5 and to fail the request if the request body size is < 5.